apt/minimization/minimizer.py

```diff
@@ -315,30 +359,58 @@ def transform(self, X):
             # values (leaves others untouched)
             if indexes and not representatives.columns.empty:
                 if len(indexes) > 1:
-                    replace = pd.concat([representatives.loc[i].to_frame().T]*len(indexes)).reset_index(drop=True)
+                    replace = pd.concat([representatives.loc[i].to_frame().T] * len(indexes)).reset_index(drop=True)
                 else:
                     replace = representatives.loc[i].to_frame().T.reset_index(drop=True)
                 replace.index = indexes
                 generalized.loc[indexes, representatives.columns] = replace
-
-        return generalized.to_numpy()
+        if self.type == 'np':
+            return generalized.to_numpy()
+        return generalized
 
     def _get_record_indexes_for_cell(self, X, cell, mapped):
-        return [i for i, x in enumerate(X) if not mapped.item(i) and
-                self._cell_contains(cell, x, i, mapped)]
+        indexes = []
+        for index, row in X.iterrows():
+            if not mapped.item(index) and self._cell_contains(cell, row, index, mapped):
+                indexes.append(index)
+        return indexes
 
     def _cell_contains(self, cell, x, i, mapped):
         for f in self._features:
             if f in cell['ranges']:
                 if not self._cell_contains_numeric(f, cell['ranges'][f], x):
                     return False
+            if f in cell['categories']:
+                if not self._cell_contains_categorical(f, cell['categories'][f], x):
+                    return False
             else:
-                #TODO: exception - feature not defined
+                # TODO: exception - feature not defined
```

Comment:
Could you please take care of the TODO by adding an appropriate exception?

Reply:
Let's make the string a little more self explanatory, something like "feature " + f + " not defined in cell " + cell['id']